Best Practices Guide: Vendor Migration (#702)
DevCycleHQ/devcycle-docs: docs/best-practices/migration.md

---
title: Migrating to DevCycle with OpenFeature

> sidebar_label: Vendor Migration

Reviewer: Is `Vendor Migration` the best term for this? I wonder if there is a clearer title for this...

Author: Good Call - Going to have the sidebar label as "Migrating to DevCycle"

description: Best Practices for Migrating to DevCycle with OpenFeature
sidebar_position: 6
sidebar_custom_props: { icon: carbon:migrate }
---
# Best Practices for Migrating to DevCycle with OpenFeature

## Introduction

Migrating from an existing feature flagging system to a new one can be complex, especially with a large codebase. OpenFeature provides a standardized interface that simplifies this process. This guide outlines best practices for leveraging OpenFeature and its Multi-Provider to ensure a smooth transition from your existing feature flagging system to DevCycle.

## Why Use OpenFeature?

### Benefits of OpenFeature

- **Standardized Interface**: Provides a consistent way to evaluate feature flags, irrespective of the underlying provider.

> - **Community Driven**": An open source CNCF incubating project under the Apache 2 license.

Reviewer: ```suggestion
- **Community Driven**: An open source CNCF incubating project under the Apache 2 license.
```

For more details, visit the [OpenFeature homepage](https://openfeature.dev) and review their [documentation](https://openfeature.dev/docs/reference/intro/).

## Best Practices for Migration

### Step 1. Abstract Evaluations

> Instead of directly calling a specific Provider throughout your code, abstract flag evaluations using OpenFeature. This abstraction makes future migrations easier between providers, or utilizing multiple providers.

Reviewer 2: ```suggestion
Instead of directly calling a specific Provider throughout your code, abstract flag evaluations using OpenFeature. This abstraction makes future migrations between providers easier, and enables the use of multiple providers at the same time.
```

```tsx
const myFlagProvider = new MyFlagProvider()
await OpenFeature.setProviderAndWait(myFlagProvider)

const openFeatureClient = OpenFeature.getClient();
const value = await openFeatureClient.getStringValue(
  'flag-key', 
  'default value', 
  someUser
);
```

### Step 2. Implement Multi-Provider

Use the Multi-Provider feature in OpenFeature to handle gradual migrations. This allows you to use multiple providers simultaneously and define rules for flag evaluation.

> #### Step 3. Setting Up Multi-Provider

Reviewer: ```suggestion
### Step 3. Setting Up Multi-Provider
```

Should this be the same level of header as Step 2? 

<img width="951" alt="image" src="https://github.com/DevCycleHQ/devcycle-docs/assets/1219069/6a56b128-de71-4f6c-b6ca-a13f0ae72cfa">

       { provider: new ExistingProvider() }
     ],
     new FirstMatchStrategy()
   );
   ```

2. **FirstSuccessfulStrategy**
Similar to `FirstMatchStrategy`, but errors from providers do not halt execution. It returns the first successful result from a provider.
   ```tsx
   const multiProvider = new MultiProvider(
     [
       { provider: new DevCycleProvider() },
       { provider: new ExistingProvider() }
     ],
     new FirstSuccessfulStrategy()
   );
   ```

3. **ComparisonStrategy**
Requires that all providers agree on a value. If they return the same value, the result is that value; otherwise, an error is returned. This is useful for validating the success of a provider migration.
   ```tsx
   const devcycleProvider = new DevCycleProvider();
   const existingProvider = new ExistingProvider();

   const onMismatch = (_resolutions: ProviderResolutionResult<FlagValue>[]) => {
     logger.warn('Provider mismatch!');
   };

   const multiProvider = new MultiProvider(
     [
       { provider: devcycleProvider },
       { provider: existingProvider }
     ],
     new ComparisonStrategy(
       existingProvider,
       onMismatch
     )
   );
   ```

### Step 4. Port Data

> Ensure all your flagging data is ported from your existing Provider to DevCycle. This involves transferring flag definitions, user targeting rules, and any associated metadata.

Reviewer 2: might want to mention that you can also do this gradually by eg. creating new flags in devcycle and keeping existing ones in old system until they are retired

Reviewer 2: can also position it as an exciting time to finally clean up all your old flags :)

> The DevCycle team has even build a [Feature Importer](https://docs.devcycle.com/integrations/feature-importer) to help with this process that is open source and customizable to import flags from either internal or 3rd party platforms.

Reviewer 2: ```suggestion
The DevCycle team has even built a [Feature Importer](https://docs.devcycle.com/integrations/feature-importer) to help with this process that is open source and can be customized to import flags from either internal or 3rd party platforms.
```

---
title: Migrating to DevCycle with OpenFeature
sidebar_label: Vendor Migration
description: Best Practices for Migrating to DevCycle with OpenFeature
sidebar_position: 6
sidebar_custom_props: { icon: carbon:migrate }
---
# Best Practices for Migrating to DevCycle with OpenFeature

## Introduction

Migrating from an existing feature flagging system to a new one can be complex, especially with a large codebase. OpenFeature provides a standardized interface that simplifies this process. This guide outlines best practices for leveraging OpenFeature and its new Multi-Provider to ensure a smooth transition from your existing feature flagging system to DevCycle.

## Why Use OpenFeature?

### Benefits of OpenFeature

- **Standardized Interface**: Provides a consistent way to evaluate feature flags, irrespective of the underlying provider.

> - **Production-Ready SDKs**: Handles configuration data fetching, error handling, and input verification.

Reviewer 3: To me this isn't a hallmark of production ready - this is what I would expect as a bare minimum.

> - **Reduced Maintenance**: Minimizes custom code required for feature flag management.

Reviewer 3: This also isn't a reduced maintenance aspect - I would lean towards how it's an open standard with backwards compatibility.

- **Easy Integration**: Supports various feature flagging systems with built-in providers and custom provider options.

## Best Practices for Migration

### Abstract Evaluations

> Instead of directly calling your Provider throughout your code, abstract flag evaluations using OpenFeature. This abstraction makes future migrations easier.

Reviewer 3: ```suggestion
Instead of directly calling a specific Provider throughout your code, abstract flag evaluations using OpenFeature. This abstraction makes future migrations easier between providers, or utilizing multiple providers.
```

```tsx
const myFlagProvider = new MyFlagProvider()
await OpenFeature.setProviderAndWait(myFlagProvider)

const openFeatureClient = OpenFeature.getClient();
const value = await openFeatureClient.getStringValue(
  'flag-key', 
  'default value', 
  someUser
);
```

### Implement Multi-Provider

Use the Multi-Provider feature in OpenFeature to handle gradual migrations. This allows you to use multiple providers simultaneously and define rules for flag evaluation.

#### Setting Up Multi-Provider

```tsx
const multiProvider = new MultiProvider([
  { provider: new DevCycleProvider() },

>   { provider: new ExsitingProvider() }

Reviewer 3: ```suggestion
  { provider: new ExistingProvider() }
```

Reviewer 3: I would generally recommend using an IDE to test code just for syntax and such

]);

await OpenFeature.setProviderAndWait(multiProvider);

const openFeatureClient = OpenFeature.getClient();
const value = await openFeatureClient.getStringValue(
  'flag-key', 
  'default value', 
  someUser
);
```

#### Strategies

The Multi-Provider supports several strategies to control how providers are evaluated and how the final result is determined.

1. **FirstMatchStrategy**

> This is the default strategy, and evaluates providers in order, moving to the next provider only if the current one returns a `FLAG_NOT_FOUND` result. This is useful for preferring DevCycle results while keeping the existing Provider as a fallback.

Reviewer 3: Might want to explain what these return types are a bit more. Unless you're familiar with OF - it's hard to understand that these are the responsibility of the provider to implement eval reasons.

Author: I borrowed heavily from @emmawillis's announcement post here - so maybe she has ideas?

Reviewer 4: @JamieSinn you mean the line about the FLAG_NOT_FOUND result? we could just say something like 'moving to the next provider only if the DevCycle provider indicates the flag doesn't exist in DevCycle'

Reviewer 3: Yeah, that works

     [
       { provider: new DevCycleProvider() },
       { provider: new ExistingProvider() }
     ],
     new FirstSuccessfulStrategy()
   );
   ```

3. **ComparisonStrategy**
Requires that all providers agree on a value. If they return the same value, the result is that value; otherwise, an error is returned. This is useful for validating the success of a provider migration.
   ```tsx
   const devcycleProvider = new DevCycleProvider();
   const existingProvider = new ExistingProvider();

   const onMismatch = (_resolutions: ProviderResolutionResult<FlagValue>[]) => {
     logger.warn('Provider mismatch!');
   };

   const multiProvider = new MultiProvider(
     [
       { provider: devcycleProvider },
       { provider: existingProvider }
     ],
     new ComparisonStrategy(
       existingProvider,
       onMismatch
     )
   );
   ```

### Port Data

Ensure all your flagging data is ported from your existing Provider to DevCycle. This involves transferring flag definitions, user targeting rules, and any associated metadata.

### Test and Validate

Before fully transitioning to DevCycle, conduct thorough testing to ensure that:

- Flag evaluations return expected results.

> - Performance is not negatively impacted.

Reviewer 3: Is this something that you could give an example of how to do via a metrics provider as a part of a multi provider setup?

Author: Potentially but I don't think it is fully implemented yet.

- There are no regressions in existing functionality.

> ### Monitor and Log

Reviewer 3: IMO this should be before the test/validate step

Implement monitoring and logging to track the performance and correctness of flag evaluations. This helps in identifying and resolving any issues during the migration.

> ### Document and Train

Reviewer 3: Same here - test and validate should be basically the last step

Author: 👍
